src/main/scala/org/monarchinitiative/dosdp/DOSDP.scala

```diff
@@ -251,3 +273,55 @@ final case class OPA(
                       edge: List[String], //TODO require length of 3
                       annotations: Option[List[Annotations]],
                       not: Option[Boolean])
+
+final case class MultiClausePrintf(
+                                   sep: Option[String],
+                                   clauses: Option[List[PrintfClause]])
+
+final case class PrintfClause(
+                               text: String,
+                               vars: Option[List[String]],
+                               sub_clauses: Option[List[MultiClausePrintf]])
+
+final case class InternalVariable(
+                                 var_name: String,
+                                 apply: Option[Function],
+                                 input: String,
+                                 )
+
+sealed trait Function {
+  def apply(input_var:Option[Binding]): Option[String]
+}
+
+object Function {
+
+  implicit val decodeFunction: Decoder[Function] = Decoder[JoinFunction].map[Function](identity).or(Decoder[RegexFunction].map[Function](identity))
+  implicit val encodeFunction: Encoder[Function] = Encoder.instance {
+    case join @ JoinFunction(_) => join.asJson
+    case regex @ RegexFunction(_) => regex.asJson
+  }
+
+}
+
+final case class JoinFunction(join: Join) extends Function {
+  override def apply(input_var:Option[Binding]): Option[String] = {
+    val joined_values = input_var.getOrElse(MultiValue(Set.empty[String])).asInstanceOf[MultiValue].value.mkString(join.sep)
+    if (joined_values.isEmpty) None else Some(joined_values)
+  }
+}
+
+final case class RegexFunction(regex: RegexSub) extends Function {
+  override def apply(input_var:Option[Binding]): Option[String] = {
+    val applied_value = input_var.getOrElse(SingleValue("")).asInstanceOf[SingleValue].value
```

Comment:
Can this be written without casting (`asInstanceOf`)?

Reply:
`asInstanceOf` based logic re-implemented with `collect { case a: SingleValue => a }`